Node/src/ethereum_l1/propose_batch_builder.rs

```
@@ -0,0 +1,310 @@
+    ) -> TransactionRequest {
+        tx.with_gas_limit(gas_limit)
+            .with_max_fee_per_gas(fees_per_gas.max_fee_per_gas)
+            .with_max_priority_fee_per_gas(fees_per_gas.max_priority_fee_per_gas)
+    }
+
+    fn update_eip4844(
+        &self,
+        tx: TransactionRequest,
+        fees_per_gas: &FeesPerGas,
+        gas_limit: u64,
+    ) -> TransactionRequest {
+        tx.with_gas_limit(gas_limit)
+            .with_max_fee_per_gas(fees_per_gas.max_fee_per_gas)
+            .with_max_priority_fee_per_gas(fees_per_gas.max_priority_fee_per_gas)
+            .with_max_fee_per_blob_gas(fees_per_gas.base_fee_per_blob_gas)
+    }
+
+    async fn get_eip1559_cost(&self, fees_per_gas: &FeesPerGas, gas_used: u64) -> u128 {
+        // Calculate the cost of the transaction based on the current base fee and priority fee.
+        // The cost is calculated as:
+        // ( (base fee + priority fee) x units of gas used).
+        // max_fee_per_gas = base_fee_per_gas + max_priority_fee_per_gas
+        (fees_per_gas.base_fee_per_gas + fees_per_gas.max_priority_fee_per_gas) * gas_used as u128
+    }
+
+    async fn get_eip4844_cost(
+        &self,
+        fees_per_gas: &FeesPerGas,
+        blob_count: u64,
+        gas_used: u64,
+    ) -> u128 {
+        let blob_gas_used = alloy::eips::eip4844::DATA_GAS_PER_BLOB * blob_count;
+        let execution_gas_cost = gas_used as u128
+            * (fees_per_gas.base_fee_per_gas + fees_per_gas.max_priority_fee_per_gas);
+        let blob_gas_cost = blob_gas_used as u128 * fees_per_gas.base_fee_per_blob_gas;
+        execution_gas_cost + blob_gas_cost
+    }
+
+    async fn get_fees_per_gas(&self) -> Result<FeesPerGas, Error> {
+        // Get base fee per gas
+        let fee_history = self
+            .provider_ws
+            .get_fee_history(2, alloy::eips::BlockNumberOrTag::Latest, &[])
+            .await?;
+
+        let base_fee_per_gas = fee_history.base_fee_per_gas.last().copied().unwrap_or(0);
```

ditto, why not return an error?